genkit-tools/common/src/eval/localFileDatasetStore.ts

```diff
@@ -0,0 +1,221 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import crypto from 'crypto';
+import fs from 'fs';
+import { appendFile, readFile, rm, writeFile } from 'fs/promises';
+import os from 'os';
+import path from 'path';
+import { v4 as uuidv4 } from 'uuid';
+import { CreateDatasetRequest, UpdateDatasetRequest } from '../types';
+import {
+  Dataset,
+  DatasetMetadata,
+  DatasetMetadataSchema,
+  DatasetStore,
+  EvalFlowInputSchema,
+} from '../types/eval';
+import { logger } from '../utils/logger';
+
+/**
+ * A local, file-based DatasetStore implementation.
+ */
+export class LocalFileDatasetStore implements DatasetStore {
+  private readonly storeRoot;
+  private readonly indexFile;
+  private readonly INDEX_DELIMITER = '\n';
+  private static cachedDatasetStore: LocalFileDatasetStore | null = null;
+
+  private constructor() {
+    this.storeRoot = this.generateRootPath();
+    this.indexFile = this.getIndexFilePath();
+    fs.mkdirSync(this.storeRoot, { recursive: true });
+    if (!fs.existsSync(this.indexFile)) {
+      fs.writeFileSync(path.resolve(this.indexFile), '');
+    }
+    logger.info(
+      `Initialized local file dataset store at root: ${this.storeRoot}`
+    );
+  }
+
+  static getDatasetStore() {
+    if (!this.cachedDatasetStore) {
+      this.cachedDatasetStore = new LocalFileDatasetStore();
+    }
+    return this.cachedDatasetStore;
+  }
+
+  static reset() {
+    this.cachedDatasetStore = null;
+  }
+
+  async createDataset(req: CreateDatasetRequest): Promise<DatasetMetadata> {
+    return this.createDatasetInternal(req.data, req.displayName);
+  }
+
+  private async createDatasetInternal(
+    data: Dataset,
+    displayName?: string
+  ): Promise<DatasetMetadata> {
+    const datasetId = this.generateDatasetId();
+    const filePath = path.resolve(
+      this.storeRoot,
+      this.generateFileName(datasetId)
+    );
+
+    if (!fs.existsSync(filePath)) {
```

Comment:
Should this be `if (fs.existsSync...)` ?

Reply:
Done.